src/callbacks_step/amr.jl

```diff
@@ -313,6 +319,115 @@ function (amr_callback::AMRCallback)(u_ode::AbstractVector, mesh::TreeMesh,
+                                     t, iter;
+                                     only_refine=false, only_coarsen=false,
+                                     passive_args=())
+  @unpack controller, adaptor = amr_callback
+
+  u = wrap_array(u_ode, mesh, equations, dg, cache)
+  lambda = @timed timer() "indicator" controller(u, mesh, equations, dg, cache,
+                                                 t=t, iter=iter)
+
+  @boundscheck begin
+    @assert axes(lambda) == (Base.OneTo(ncells(mesh)),) (
+      "Indicator array (axes = $(axes(lambda))) and mesh cells (axes = $(Base.OneTo(ncells(mesh)))) have different axes"
+    )
+  end
+
+  # Copy controller value of each quad to the quad's user data storage
+  iter_volume_c = @cfunction(copy_to_quad_iter_volume, Cvoid, (Ptr{p4est_iter_volume_info_t}, Ptr{Cvoid}))
+
+  # The pointer to lambda will be interpreted as Ptr{Int} above
+  @assert lambda isa Vector{Int}
+
+  GC.@preserve lambda begin
+    p4est_iterate(mesh.p4est,
+                  C_NULL, # ghost layer
+                  pointer(lambda),
+                  iter_volume_c, # iter_volume
+                  C_NULL, # iter_face
+                  C_NULL) # iter_corner
+  end
+
+  @timed timer() "refine" if !only_coarsen
+    # Refine mesh
+    refined_original_cells = @timed timer() "mesh" refine!(mesh)
+
+    # Refine solver
+    @timed timer() "solver" refine!(u_ode, adaptor, mesh, equations, dg, cache,
+                                    refined_original_cells)
+    for (p_u_ode, p_mesh, p_equations, p_dg, p_cache) in passive_args
+      @timed timer() "passive solver" refine!(p_u_ode, adaptor, p_mesh, p_equations,
+                                              p_dg, p_cache, refined_original_cells)
+    end
+  else
+    # If there is nothing to refine, create empty array for later use
+    refined_original_cells = Int[]
+  end
+
+  @timed timer() "coarsen" if !only_refine
+    # Coarsen mesh
+    coarsened_original_cells = @timed timer() "mesh" coarsen!(mesh)
+
+    # coarsen solver
+    @timed timer() "solver" coarsen!(u_ode, adaptor, mesh, equations, dg, cache,
+                                     coarsened_original_cells)
+    for (p_u_ode, p_mesh, p_equations, p_dg, p_cache) in passive_args
+      @timed timer() "passive solver" coarsen!(p_u_ode, adaptor, p_mesh, p_equations,
+                                               p_dg, p_cache, coarsened_original_cells)
+    end
+  else
+    # If there is nothing to coarsen, create empty array for later use
+    coarsened_original_cells = Int[]
+  end
+
+  # Store whether there were any cells coarsened or refined
+  has_changed = !isempty(refined_original_cells) || !isempty(coarsened_original_cells)
+  if has_changed # TODO: Taal decide, where shall we set this?
+    # don't set it to has_changed since there can be changes from earlier calls
+    mesh.unsaved_changes = true
+
+    reinitialize_boundaries!(semi.boundary_conditions, cache)
```

Comment:
@efaulhaber I am wondering - why is this invoked here? It seems to be out of place, something so very specific to be called here in this "general" AMR routine. Why isn't it, for example, part of the call to `reinitialize_containers!`? Wouldn't that be a better fit?

Reply:
Would you prefer passing the semidiscretization to the "refine/coarsen solver" functions, and from there to `reinitialize_containers!`?

Reply:
Why the semi? If I recall our earlier discussion correctly, the problem was the missing cache, which is already passed here? Or was this a different issue?

Reply:
The problem was the boundary conditions container. This is stored is `semi.boundary_conditions`, so I need to pass either the semi or the boundary conditions to the "refine/coarsen solver" functions (and to `reinitialize_containers!`).

Reply:
OK. I guess it's ok to leave it as it is here, and maybe we'll figure out a more canonical way how to handle these kind of changes in the future.